### ISSUE TITLE
Getting columns back but no data

### ISSUE THREAD

**Issue by jayhawkins** [mevdschee/php-crud-api#243]
Quirky issue I'm running into. Doing a GET for a record on my dev box and everything works great. Pushed it to testing/prod server and I get back the columns only, but no data. Weird thing is everything is fine via Postman too. It only seems to be on my test/prod server, so figuring it's a configuration or compile issue, just not sure if you've seen it before and can point me in the right direction?

    _site.com_/api/email_templates?filter=title,eq,Authorize Account

stdClass Object ( [email_templates] => stdClass Object ( [columns] => Array ( [0] => id [1] => title [2] => body [3] => entityID [4] => createdAt [5] => updatedAt [6] => subject ) [records] => Array ( ) ) ) Template Result: 1

**Comment by mevdschee** [comment]
>Pushed it to testing/prod server and I get back the columns only, but no data.

Do you run the latest version? What versions of PHP/MySQL do you run on testing/prod?

**Comment by jayhawkins** [comment]
There are no errors. I get a 200 back, although at this point I'm die()'ing so I can manually dump my array. But previous investigations into logs were showing no errors. The only errors I came across were when trying to access a records element in my php code and it failed with invalid offset 0 (because there were columns returned, but no records.

Dev(works) - Debian 8 / PHP 7.1 / MySQL 5.7.17
Test(fails) - Ubuntu 16.10 / PHP 7.0 / MySQL 5.7.17

Latest version of Ubuntu might worry me a little?

See if you can hit this:   http://nec.dubtel.com/info.php

**Comment by mevdschee** [comment]
@jayhawkins There should not be any problem with those version. Can you download the api.php file from the repo and do a "diff" to see what may have changed?

**Comment by jayhawkins** [comment]
Using Postman - 

This is the result dev gets:
http://api.nec.dubtel.com/api/email_templates?filter=title,eq,Authorize Account

If you wrap the Authorize Account in quotes, it is the result test gets:
http://api.nec.dubtel.com/api/email_templates?filter=title,eq,'Authorize Account'

I wonder if it has to do with how apache may be accepting the query strings? My dev box is running Nginx, the test box is running Apache2, sorry I forgot to mention that.

Jay

**Comment by jayhawkins** [comment]
This is the PHP code that is running and returning the empty records array. Basically, the same thing I'm doing with Postman which gives back records array data:

$templateurl = API_HOST."/api/email_templates?filter=title,eq,Authorize Account";
                        $templateoptions = array(
                            'http' => array(
                                'header'  => "Content-type: application/x-www-form-urlencoded\r\n",
                                'method'  => 'GET'
                            )
                        );
                        $templatecontext  = stream_context_create($templateoptions);
                        $templateresult = json_decode(file_get_contents($templateurl,false,$templatecontext));
                        echo "Template Result: " . print_r($templateresult);
                        die();

**Comment by jayhawkins** [comment]
I downloaded the latest file and did a diff on what I have. The only change is line 201, which is your changes for binary and varbinary. I wouldn't believe that to be the issue. 

Since Postman works fine, I am under the assumption it may be PHP. But it's still early and I'm only guessing.

**Comment by mevdschee** [comment]
Use PHP `urlencode()` when using non-URL safe characters (like space). Space is represented as %20. Your code should have:

    $templateurl = API_HOST."/api/email_templates?filter=title,eq,".urlencode("Authorize Account");

or simply:

    $templateurl = API_HOST."/api/email_templates?filter=title,eq,Authorize%20Account";

I think that should do it.

**Comment by mevdschee** [comment]
Nginx is more forgiving when you send it invalid URLs and does this conversion for you internally (Apache does not). See also:

http://php.net/manual/en/function.http-build-query.php

I would recommend to build the string like this:

    $templateargs = array("filter"=>"title,eq,Authorize Account");
    $templateurl = API_HOST."/api/email_templates?".http_build_query($templateargs);

I'm pretty sure that this is the issue you are running into.

**Comment by jayhawkins** [comment]
Resolved!

Thanks for the insight. I was wrapping the whole templateurl in the urlencode. I used your suggestion of adding the $templateargs array. It fit in with how I was doing my POST requests so everything is more uniform.

Also wanted to add that in my .htaccess file I had to use:
RewriteRule ^api/(.*)$ api.php/$1 [L,QSA]
to get it working. Just wanted you to know.

Much appreciated. Your script(s) will make our mobile front end much easier to implement also!
